gtsfm/pose_slam/pose_slam.py

```diff
@@ -20,6 +23,100 @@
+        """Calculate the translation and angle differences of two poses.
+        P1, P2: Pose3
+        Return:
+            distance: translation difference
+            angle: angular difference
+        """
+        # TODO(frank): clean this up
+        t1 = P1.translation()
+        t2 = P2.translation()
+        R1 = P1.rotation()
+        R2 = P2.rotation()
+        R1_2 = R1.compose(R2.inverse())
+        t1_ = R1_2.rotate(t2)
+        # t1_2 = t1 - R1_2*t2
+        distance = np.linalg.norm(t1 - t1_)
+        angle_ = PoseSlam.angle(R1, R2)
+        return distance, angle_
+
+    @staticmethod
+    def filter_constraints(constraints: List[Constraint], poses: List[Pose3]) -> List[Constraint]:
+        """Filter constraints according to notebook criteria.
+
+        Args:
+            constraints (List[Constraint]): relative constraints from file
+            poses (List[Pose3]): estimated initial values
+
+        Returns:
+            Filtered constraints.
+        """
+
+        filtered_constraints: List[Constraint] = []
+
+        # Filter on covariance and error
+        for constraint in constraints:
+            a, b = constraint.a, constraint.b
+            aTb, cov = constraint.aTb, constraint.cov
+            predicted_aTb = poses[a].between(poses[b])
+            trans_diff, rot_diff = PoseSlam.difference(aTb, predicted_aTb)
+            inlier = (trans_diff <= 0.04) and (rot_diff <= 5)
+            if not inlier:
+                continue
+            try:
+                info = np.linalg.inv(cov)
+                if np.isnan(info).any():
+                    continue
+                else:
+                    filtered_constraints.append(constraint)
+            except np.linalg.LinAlgError:
+                continue
+
+        return filtered_constraints
+
+    @staticmethod
+    def filtered_pose_priors(
+        constraints: List[Constraint], poses: List[Pose3], add_backbone=True
+    ) -> Dict[Tuple[int, int], PosePrior]:
+        """Generate relative pose priors from constraints and initial_estimate by filtering heavily.
+
+        Args:
+            constraints (List[Constraint]): relative constraints from file
+            poses (List[Pose3]): estimated initial values
+            add_backbone (bool, optional): Add soft backbone constraints. Defaults to True.
+
+        Returns:
+            Dict[Tuple[int, int], PosePrior]: dictionary of relative pose priors.
+        """
+
+        relative_pose_priors: Dict[Tuple[int, int], PosePrior] = {}
+        for constraint in PoseSlam.filter_constraints(constraints, poses):
+            a, b = constraint.a, constraint.b
+            aTb, cov = constraint.aTb, constraint.cov
+            relative_pose_priors[(a, b)] = PosePrior(aTb, cov, PosePriorType.SOFT_CONSTRAINT)
+
+        # Create loose odometry factors for backbone.
+        if add_backbone:
+            backbone_cov = np.diag(np.array([1, 1, 1, np.deg2rad(30.0), np.deg2rad(30.0), np.deg2rad(30.0)]))
```

Comment:
If this is correct (rotations after translations in the diagonal), we have a bug in 1DSFM: https://github.com/borglab/gtsfm/blob/hkust-cluster/gtsfm/averaging/translation/averaging_1dsfm.py#L184?

Reply:
I was going with this comment, translations after rotations in the tangent space.  https://github.com/borglab/gtsam/blob/da00870ee1a7ce8fe746008abde62ad012dd177b/gtsam/geometry/Pose3.h#L144